kernel/mouse/src/lib.rs

```diff
@@ -56,18 +100,20 @@ pub fn init(mut mouse: PS2Mouse<'static>, mouse_queue_producer: Queue<Event>) ->
 /// 
 /// In some cases (e.g. on device init), [the PS/2 controller can also send an interrupt](https://wiki.osdev.org/%228042%22_PS/2_Controller#Interrupts).
 extern "x86-interrupt" fn ps2_mouse_handler(_stack_frame: InterruptStackFrame) {
-    if let Some(MouseInterruptParams { mouse, queue }) = MOUSE.get() {
+    if let Some(MouseInterruptParams { mouse, queue, packet_bytes }) = MOUSE.get() {
+        // using `while` here didn't interact well with the window manager and increases handler runtime
         if mouse.is_output_buffer_full() {
-            // NOTE: having read some more forum comments now, if this ever breaks on real hardware,
-            // try to redesign this to only get one byte per interrupt instead of the 3-4 bytes we
-            // currently get in read_mouse_packet and merge them afterwards
-            let mouse_packet = mouse.read_mouse_packet();
-            
-            if mouse_packet.always_one() != 1 {
-                // this could signal a hardware error or a mouse which doesn't conform to the rule
-                warn!("ps2_mouse_handler(): Discarding mouse data packet since its third bit should always be 1.");
-            } else if let Err(e) = handle_mouse_input(mouse_packet, queue) {
-                error!("ps2_mouse_handler(): {e:?}");
+            packet_bytes.push(mouse.read_packet_byte());
+
+            if let Some(bytes) = packet_bytes.filled_bytes() {
```

Comment:
these two instructions acquire and release a total of 4 locks. You can likely combine their functionality together such that a "push" operation returns an `Option<[u8; N]>` indicating whether the expected number of bytes have been received (and providing a copy of those 3 or 4 bytes).

Also, if you adopt a single-Mutex design (or better yet, the AtomicU64 design), that would reduce this overhead from 4 to 1 lock acquisitions.

Reply:
Again, thanks, exactly what I need to learn.

Reply:
yea np at all, this is just the kind of thing that only becomes apparent once you have another set of eyes on it. :smile: